// testcafe-browser-provider-browserstack/lib/backends/base.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ERROR_MESSAGES = __importStar(require("../templates/error-messages"));
class BaseBackend {
    constructor(reportWarning) {
        this.reportWarning = reportWarning;
    }
    async getBrowsersList() {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async openBrowser( /*id, pageUrl, capabilities*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async closeBrowser( /*id*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async takeScreenshot( /*id, path*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async resizeWindow( /*id, width, height, currentWidth, currentHeight*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async maximizeWindow( /*id*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async reportJobResult( /*id, jobStatus, jobData, possibleStatuses*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    getSessionUrl( /*id*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
    async getOSInfo( /*id*/) {
        throw new Error(ERROR_MESSAGES.API_METHOD_NOT_IMPLEMENTED());
    }
}
exports.default = BaseBackend;
