// testcafe-browser-provider-browserstack/lib/browser-proxy.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const http_1 = __importDefault(require("http"));
const url_1 = require("url");
const pinkie_1 = __importDefault(require("pinkie"));
module.exports = class BrowserProxy {
    constructor(targetHost, targetPort, { proxyPort, responseDelay } = {}) {
        this.targetHost = targetHost;
        this.targetPort = targetPort;
        this.proxyPort = proxyPort || 0;
        this.responseDelay = responseDelay || 0;
        this.server = http_1.default.createServer((...args) => this._onBrowserRequest(...args));
        this.server.on('connection', socket => socket.unref());
    }
    _onBrowserRequest(req, res) {
        setTimeout(() => {
            const parsedRequestUrl = (0, url_1.parse)(req.url);
            const destinationUrl = 'http://' + this.targetHost + ':' + this.targetPort + parsedRequestUrl.path;
            res.statusCode = 302;
            res.setHeader('location', destinationUrl);
            res.end();
        }, this.responseDelay);
    }
    async init() {
        return new pinkie_1.default((resolve, reject) => {
            this.server.listen(this.proxyPort, err => {
                if (err)
                    reject(err);
                else {
                    this.proxyPort = this.server.address().port;
                    resolve();
                }
            });
        });
    }
    dispose() {
        this.server.close();
    }
};
